10286: UNIXAuthorizedKeysFiles type issues

twm opened the pull request:
## Scope and purpose


It turns out all the tests for this code were skipped because the `skip` values were the empty string, rather than ``None``, when the conditional imports succeeded. I've rearranged the imports a bit so that the tests run, and made them typecheck with the changes from #1676.

The actual fix here is pretty simple: decode the username [as in the fix to #9130](https://twistedmatrix.com/trac/ticket/9130). As usual the stdlib's obsession with making everything `str` instead of `bytes` ~bites~ costs us in pointless copies and additional error cases. 🙄

Getting it all to type check... well, I'm hoping that #1677 will allow me to remove some of the casts here, as [typeshed has typings for `pwd`](https://github.com/python/typeshed/blob/master/stdlib/pwd.pyi) and [`spwd`](https://github.com/python/typeshed/blob/master/stdlib/spwd.pyi).

The layering here isn't great — the implementation mixes a pluggable `pwd` backend with the system-level concern of filesystem encoding — but I don't see a _simple_ fix that preserves backwards compatibility. Easier to replace ``UNIXAuthorizedKeysFiles`` entirely, which I've no ambition for.

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10286
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [x] I have updated the automated tests and checked that all checks for the PR are green.
* [ ] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

twm committed:
Fix type errors in test_checkers

twm committed:
Fix erroneous skips

The runner skips on "". skip must be None to not skip.

twm committed:
Assert synchronously

twm committed:
Get tests passing

twm committed:
Clean up types

twm committed:
Remove unnecessary type: ignore

twm committed:
Add newsfragment

twm committed:
Wrap long line

exarkun commented on src/twisted/conch/newsfragments/10286.bugfix:
```diff
@@ -0,0 +1 @@
+twisted.conc.checkers.UNIXAuthorizedKeysFiles now uses the filesystem encoding to decode usernames before looking them up in the password database, so it works on CPython 3.
```
type in the module name here.  also, pypy is still supported I think?  So maybe Python 3 instead of CPython 3.

twm replied:
It actually already worked on PyPy 3 due to different (IMO better) behavior there, but this is perhaps more confusing than illuminating to mention here.

adiroiban replied:
```suggestion
twisted.conch.checkers.UNIXAuthorizedKeysFiles now uses the filesystem encoding to decode usernames before looking them up in the password database.
This is a fix for CPython 3. PyPy3 was not affected by this bug.
```

adiroiban commented on src/twisted/conch/checkers.py:
```diff
@@ -47,34 +44,115 @@
     if pwd is None:
         return None
-    return pwd.getpwnam(username)
+    return cast(CryptedPasswordRecord, pwd.getpwnam(username))
```
#1677 is merged. Can this cast be removed?

twm replied:
Unfortunately it seems not, as MyPy still complains:

```
src/twisted/conch/checkers.py:155:5: error: Returning Any from function declared to return "Optional[CryptedPasswordRecord]" 
[no-any-return]
        return runAsEffectiveUser(0, 0, f, username)
```

adiroiban commented on src/twisted/conch/checkers.py:
```diff
@@ -47,34 +44,115 @@
     else:
         return None
-    return runAsEffectiveUser(0, 0, f, username)
+    return cast(CryptedPasswordRecord, runAsEffectiveUser(0, 0, f, username))
```
is this cast still needed after #1677?

twm replied:
Same story:

```
src/twisted/conch/checkers.py:134:12: error: Incompatible return value type (got "UserRecord", expected
"Optional[CryptedPasswordRecord]")  [return-value]
        return pwd.getpwnam(username)
```

adiroiban commented on src/twisted/conch/checkers.py:
```diff
@@ -47,34 +44,115 @@
+
+    @raises KeyError: when the user doesn't exist
+    """
+    return userdb.getpwnam(username.decode(sys.getfilesystemencoding()))
```
This looks OK as I see a similar code in `UNIXPasswordDatabase.requestAvatarId`

I was expecting the credential username and password to be `str`.
Who is using random bytes for this?

Or allow the whole Twisted cred framework to be setup with a custom encoding.

--------

I am not using the Twisted Cred framework myself.
I think that when I tried to use it 10 years ago, when I was just learning Twisted I ended up with lots of issued due to username encodings.
I had automated tests running for various encoding and at that time I didn't understood well how default encoding works and how to mock it.
...so I ended up with my custom credential code.

twm replied:
The code it's actually inverting is this:

https://github.com/python/cpython/blob/a5451c96a14ac0c3ee7cef7b5c52ab1d783ec613/Modules/pwdmodule.c#L222-L226

As I grumbled on the PyPy bug I think it's weird to force us to deal in Unicode when the actual API is `bytes` with an undefined encoding, but :man_shrugging:.

adiroiban replied:
Yes. I guess that in cpython3 due to the impulse to have everything as unicode, we ended up with this API.

This is life :)

adiroiban reviewed (approved):
It looks good. Minor comments inline.

Please check the comments and can merge once all tests are green.

Many thanks!

I have updated the branch from UI to remove 3.6 and get the tests green.

twm committed:
Merge branch 'trunk' into 10286-pwd-checkers-types

twm committed:
Address review feedback on newsfragment

twm commented:
I'll wait for https://twistedmatrix.com/trac/ticket/10304 for `Protocol`.

adiroiban commented:
Thanks for the update. Yes. Once the CI is green please merge.
You can also enable auto-merge :)

twm committed:
Merge branch 'trunk' into 10286-pwd-checkers-types

twm committed:
Import Protocol from typing_extensions

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

twm committed:
Skip SSHCheckerTests on Windows

The Windows builds are failing with this message:

    [ERROR]
    Traceback (most recent call last):
      File "D:\a\1\s\.tox\alldeps-withcov-windows\lib\site-packages\twisted\cred\test\test_strcred.py", line 434, in test_isChecker
        sshChecker = strcred.makeChecker("sshkey")
      File "D:\a\1\s\.tox\alldeps-withcov-windows\lib\site-packages\twisted\cred\strcred.py", line 117, in makeChecker
        return findCheckerFactory(authType).generateChecker(argstring)
      File "D:\a\1\s\.tox\alldeps-withcov-windows\lib\site-packages\twisted\plugins\cred_sshkeys.py", line 42, in generateChecker
        return SSHPublicKeyChecker(UNIXAuthorizedKeysFiles())
      File "D:\a\1\s\.tox\alldeps-withcov-windows\lib\site-packages\twisted\conch\checkers.py", line 527, in __init__
        raise ValueError("No pwd module found, and no userdb argument passed.")
    builtins.ValueError: No pwd module found, and no userdb argument passed.

    twisted.cred.test.test_strcred.SSHCheckerTests.test_isChecker

This exception was introduced in this PR. Previous to this PR the
UNIXAuthorizedKeysFiles initializer would allow self.userdb to be None
and generate an AttributeError later when attempting to get authorized
keys.
